### PR TITLE
stats: unrevert and fix histograms

### PR DESCRIPTION
**Pull request by mattklein123**
This change unreverts:
https://github.com/envoyproxy/envoy/pull/3130
https://github.com/envoyproxy/envoy/pull/3183
https://github.com/envoyproxy/envoy/pull/3219
Also fixes https://github.com/envoyproxy/envoy/issues/3223

Please see the 2nd commit for the actual changes. The changes are:
1) Bring in a new histogram library version with the fix (and more debug
 checking).
2) Fix a small issue with scope iteration during merging.
3) Remove de-dup for histograms until we iterate to shared
storage for overlapping scope histograms. Personally, I found
this very confusing when debugging and I think the way I changed
it is better for now given the code we have.

### CODE REVIEW

**Comment by mattklein123** [comment]
@ramaraochavali @jmarantz @mrice32 PTAL.

Unfortunately, I have some kind of HW issue with my computer and it's randomly crashing under load. I won't be able to get a replacement until Monday morning when I am back from Europe so if there are major changes required to this they are going to have to wait.

File: source/common/stats/thread_local_store.cc [hunk 1]
```diff
@@ -105,9 +101,8 @@ void ThreadLocalStoreImpl::mergeHistograms(PostMergeCb merge_complete_cb) {
     merge_in_progress_ = true;
     tls_->runOnAllThreads(
         [this]() -> void {
-          for (ScopeImpl* scope : scopes_) {
-            for (const auto& name_histogram_pair :
-                 tls_->getTyped<TlsCache>().scope_cache_[scope].histograms_) {
+          for (const auto& scopes : tls_->getTyped<TlsCache>().scope_cache_) {
```
Comment (jmarantz) [inline source/common/stats/thread_local_store.cc]: scopes or scope?  Previously it was 'scope' with a concrete type. WDYT of using a concrete type for this one as it's not super-obvious locally what type this resolves to, and I suspect it's not particularly verbose.

Comment (mattklein123) [reply source/common/stats/thread_local_store.cc]: Sure will do. Caveat being that my computer is Fd and I'm having trouble doing anything real so probably will need to wait till Monday.

Comment (jmarantz) [reply source/common/stats/thread_local_store.cc]: OK go ahead and merge then...I can do these myself easily after you merge.  I know we're trying to try these out so no need to delay a week for a  style nit.

File: source/common/stats/thread_local_store.cc [hunk 2]
```diff
@@ -208,31 +255,143 @@ Histogram& ThreadLocalStoreImpl::ScopeImpl::histogram(const std::string& name) {
...
+  TlsHistogramSharedPtr* tls_ref = nullptr;
+  if (!parent_.shutting_down_ && parent_.tls_) {
+    tls_ref = &parent_.tls_->getTyped<TlsCache>().scope_cache_[this->scope_id_].histograms_[name];
+  }
+
+  if (tls_ref && *tls_ref) {
+    return **tls_ref;
+  }
+
+  std::unique_lock<std::mutex> lock(parent_.lock_);
```
Comment (mrice32) [inline source/common/stats/thread_local_store.cc]: I don't think we need to take this lock here.  Kinda unrelated side note: I think we can make the `getTagsForName` method on `ThreadLocalStore` `const`.

Comment (mattklein123) [reply source/common/stats/thread_local_store.cc]: Yup that's right. I think I also noticed that at some point and then forgot about it amidst all the issues. Thanks for mentioning it. I will see if I can nurse my computer into making some of these small changes.

Comment (ramaraochavali) [reply source/common/stats/thread_local_store.cc]: @mattklein123 if you have trouble with your computer, merge this PR and I can do the follow-up PR addressing these minor comments if that’s ok.

Comment (mattklein123) [reply source/common/stats/thread_local_store.cc]: Sure, that's fine. @htuch can you approve/merge if this looks good?

**Pull request merged**